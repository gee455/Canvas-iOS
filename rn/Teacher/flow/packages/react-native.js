//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

declare var console: typeof console & { disableYellowBox: boolean }

export type PanResponderGestureState = {
  stateID: number,
  moveX: number,
  moveY: number,
  x0: number,
  y0: number,
  dx: number,
  dy: number,
  vx: number,
  vy: number,
}

export type KeyboardEventData = {
  endCoordinates: {
    width: number,
    height: number,
    screenX: number,
    screenY: number,
  },
}
