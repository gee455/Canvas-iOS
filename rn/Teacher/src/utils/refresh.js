//
// Copyright (C) 2016-present Instructure, Inc.
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, version 3 of the License.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
//

// @flow

import React, { Component } from 'react'
import hoistStatics from 'hoist-non-react-statics'

export type RefreshFunction = (*) => *
export type ShouldRefresh = (*) => boolean
export type IsFetchingData = (*) => boolean
export type TtlKeyExtractor = (*) => string

const DEFAULT_TTL_KEY = 'default'

export default function refresh (
  refreshFunction: RefreshFunction,
  shouldRefresh: ShouldRefresh,
  isFetchingData: IsFetchingData,
  ttl: ?number = 1000 * 60 * 60, // 1 hour
  ttlKeyExtractor: TtlKeyExtractor = (props) => DEFAULT_TTL_KEY,
) {
  let updates = {}

  return function (TheirComponent: React$ElementType) {
    class Refreshed extends Component<*, RefreshState> {
      state: RefreshState = { refreshing: false }

      refreshFunction = refreshFunction
      shouldRefresh = shouldRefresh
      isFetchingData = isFetchingData

      componentWillReceiveProps (nextProps) {
        this.setState({
          refreshing: this.state.refreshing
            ? isFetchingData(nextProps)
            : false,
        })
      }

      componentWillMount () {
        if (this.props.forceRefresh || shouldRefresh(this.props)) {
          this.setLastUpdate(Date.now())
          refreshFunction(this.props)
        } else if (!this.getLastUpdate() || Date.now() > this.getLastUpdate() + ttl) {
          this.refresh()
        }
      }

      refresh = () => {
        this.setState({ refreshing: true }, () => {
          this.setLastUpdate(Date.now())
          refreshFunction(this.props)
        })
      }

      setLastUpdate = (date: number) => {
        updates[ttlKeyExtractor(this.props)] = date
      }

      getLastUpdate = () => updates[ttlKeyExtractor(this.props)]

      render = () => <TheirComponent {...this.props} refresh={this.refresh} refreshing={this.state.refreshing} />
    }

    return hoistStatics(Refreshed, TheirComponent)
  }
}
