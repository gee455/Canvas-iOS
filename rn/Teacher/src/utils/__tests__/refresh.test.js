//
// Copyright (C) 2016-present Instructure, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//

// @flow

import React, { Component } from 'react'
import { Text } from 'react-native'
import refresh from '../refresh'
import renderer from 'react-test-renderer'
import setProps from '../../../test/helpers/setProps'

describe('refresh', () => {
  it('renders the refreshed component', () => {
    let Refreshed = refresh(() => {}, () => true, () => false)(Text)
    let tree = renderer.create(
      <Refreshed testID='test'>This is text</Refreshed>
    ).toJSON()

    expect(tree).toMatchSnapshot()
  })

  it('has all the correct statics', () => {
    class C extends Component<{}> {
      static yo = 'asdf'
    }
    let Refreshed = refresh(() => {}, () => true, () => false)(C)

    expect(Refreshed.yo).toEqual('asdf')
  })

  it('calls the refresh function if should refresh returns true', () => {
    let refreshFunction = jest.fn()
    let Refreshed = refresh(refreshFunction, () => true, () => false)(Text)
    renderer.create(
      <Refreshed>This is text</Refreshed>
    )
    expect(refreshFunction).toHaveBeenCalled()
  })

  it('calls the refresh function if forceRefresh is true', () => {
    let refreshFunction = jest.fn()
    let Refreshed = refresh(refreshFunction, () => false, () => false)(Text)
    renderer.create(
      <Refreshed forceRefresh={true}>This is text</Refreshed>
    )
    expect(refreshFunction).toHaveBeenCalled()
  })

  it('doesnt call the refresh function if should refresh return false', () => {
    let refreshFunction = jest.fn()
    let Refreshed = refresh(refreshFunction, () => false, () => false)(Text)
    let tree = renderer.create(
      <Refreshed>This is text</Refreshed>
    )
    jest.resetAllMocks()
    tree.update(<Refreshed>This is text</Refreshed>)
    expect(refreshFunction).not.toHaveBeenCalled()
  })

  it('refreshes if should refresh returns false but the ttl is expired', () => {
    let refreshFunction = jest.fn()
    let Refreshed = refresh(refreshFunction, () => false, () => false, 0)(Text)
    renderer.create(
      <Refreshed>This is text</Refreshed>
    )
    expect(refreshFunction).toHaveBeenCalled()
  })

  it('passes in the refresh function as a prop to the underlying component', () => {
    let refreshFunction = jest.fn()
    let Refreshed = refresh(refreshFunction, () => true, () => false)(Text)
    let tree = renderer.create(
      <Refreshed>This is text</Refreshed>
    ).toJSON()
    tree.props.refresh()
    expect(refreshFunction).toHaveBeenCalled()
  })

  it('can be refreshed', async () => {
    let pending = 0
    let fetchData = jest.fn((props) => {
      pending = 1
    })
    let Refreshed = refresh(fetchData, () => false, () => Boolean(pending))(Text)

    let component = renderer.create(
      <Refreshed>Some moar text</Refreshed>
    )

    let tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    tree.props.refresh()
    expect(fetchData).toHaveBeenCalled()

    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    setProps(component, { pending: 0 })
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()

    component.getInstance().setState({
      refreshing: false,
    })
    setProps(component, { pending: 0 })
    tree = component.toJSON()
    expect(tree).toMatchSnapshot()
  })
})
