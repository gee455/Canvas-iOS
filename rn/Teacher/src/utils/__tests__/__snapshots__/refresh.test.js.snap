// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`refresh can be refreshed 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  refresh={[Function]}
  refreshing={true}
>
  Some moar text
</Text>
`;

exports[`refresh can be refreshed 2`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  refresh={[Function]}
  refreshing={true}
>
  Some moar text
</Text>
`;

exports[`refresh can be refreshed 3`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  refresh={[Function]}
  refreshing={true}
>
  Some moar text
</Text>
`;

exports[`refresh can be refreshed 4`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  refresh={[Function]}
  refreshing={false}
>
  Some moar text
</Text>
`;

exports[`refresh renders the refreshed component 1`] = `
<Text
  accessible={true}
  allowFontScaling={true}
  ellipsizeMode="tail"
  refresh={[Function]}
  refreshing={false}
  testID="test"
>
  This is text
</Text>
`;
